app/validators/mac_address_validator.rb

```diff
@@ -6,7 +6,10 @@ def validate_each(record, attribute, value)
   end
 
   def make_invalid(record, attribute)
-    record.errors.add(attribute, (options[:message] || _("is not a valid MAC address")))
+    # error message can already be present from the Net::Validations::normalize_mac method
+    if record.errors[:mac].empty?
```

Comment:
Shouldn't this be for the same attribute?
```suggestion
    if record.errors[attribute].empty?
```

Reply:
Sorry for the double comment, but I just looked and in the other place it's also hardcoded:
https://github.com/theforeman/foreman/blob/2d5bd87220020f7e7a9a6c41dc2907061c22d8ee/app/services/ipam/base.rb#L64-L69

That makes me question why we call normalize twice. I'm not sure the service should add an error in the first place but it was introduced in 990dee696f58b9fa6035b239152223b073999488 so I suspect this bug has been around for a long time.

What I suspect happens is that it calls `before_validation`:
https://github.com/theforeman/foreman/blob/2d5bd87220020f7e7a9a6c41dc2907061c22d8ee/app/models/nic/base.rb#L14

https://github.com/theforeman/foreman/blob/2d5bd87220020f7e7a9a6c41dc2907061c22d8ee/app/models/nic/base.rb#L282-L287

And then it validates (which is the code you're editing now):
https://github.com/theforeman/foreman/blob/2d5bd87220020f7e7a9a6c41dc2907061c22d8ee/app/models/nic/base.rb#L24

It makes me think your current patch is a correct short term fix (because it does make still ensure it's invalid), but we should really untangle the whole mess.

Reply:
fixed.

Reply:
> Sorry for the double comment, but I just looked and in the other place it's also hardcoded:
> 
> https://github.com/theforeman/foreman/blob/2d5bd87220020f7e7a9a6c41dc2907061c22d8ee/app/services/ipam/base.rb#L64-L69
> 
> That makes me question why we call normalize twice. I'm not sure the service should add an error in the first place but it was introduced in [990dee6](https://github.com/theforeman/foreman/commit/990dee696f58b9fa6035b239152223b073999488) so I suspect this bug has been around for a long time.

I am not sure if normalize_mac and MacAddressValidator do the same thing, so I'd prefer not to touch that :smile:

Reply:
@ekohl should I revert this back to record.errors.[:mac]?